iquip/apps/explorer.py

```diff
@@ -1,32 +1,94 @@
+from typing import Optional, Tuple, Union
 
-from PyQt5.QtCore import QObject
-from PyQt5.QtWidgets import QLabel, QVBoxLayout, QWidget
+from PyQt5.QtCore import QObject, pyqtSlot
+from PyQt5.QtWidgets import (
+    QPushButton, QTreeWidget, QTreeWidgetItem, QVBoxLayout, QWidget
+)
 
 import qiwis
 
+from iquip import cmdtools
+
 class ExplorerFrame(QWidget):  # pylint: disable=too-few-public-methods
     """Frame for showing the experiment list and opening an experiment."""
 
     def __init__(self, parent: Optional[QWidget] = None):
         """Extended."""
         super().__init__(parent=parent)
         # widgets
-        self.testLabel = QLabel("explorer", self)
+        self.expStructure = QTreeWidget(self)
+        self.expStructure.header().setVisible(False)
+        self.reloadButton = QPushButton("Reload", self)
+        self.openButton = QPushButton("Open", self)
         # layout
         layout = QVBoxLayout(self)
-        layout.addWidget(self.testLabel)
+        layout.addWidget(self.reloadButton)
+        layout.addWidget(self.expStructure)
+        layout.addWidget(self.openButton)
 
 
 class ExplorerApp(qiwis.BaseApp):  # pylint: disable=too-few-public-methods
     """App for showing the experiment list and opening an experiment."""
 
-    def __init__(self, name: str, parent: Optional[QObject] = None):
+    def __init__(self, name: str, masterPath: str = ".", parent: Optional[QObject] = None):
         """Extended."""
         super().__init__(name, parent=parent)
+        self.repositoryPath = os.path.join(masterPath, "repository")
         self.explorerFrame = ExplorerFrame()
+        self.loadExpStructure()
+        # connect signals to slots
+        self.explorerFrame.reloadButton.clicked.connect(self.loadExpStructure)
+        self.explorerFrame.openButton.clicked.connect(self.openExp)
+
+    @pyqtSlot()
+    def loadExpStructure(self):
+        """Loads the experiment file structure in self.explorerFrame.expStructure.
+
+        It assumes that all experiment files are in self.repositoryPath.
+        """
+        self.explorerFrame.expStructure.clear()
+        self._addExpFile(self.repositoryPath, self.explorerFrame.expStructure)
+
+    def _addExpFile(self, path: str, parent: Union[QTreeWidget, QTreeWidgetItem]):
+        """Searches the sub elements and add them into self.explorerFrame.expStructure.
+
+        This uses the DFS algorithm.
+        1. Fetch the command result, which is a list of sub elements.
+        2. If a directory, call _addExpFile() recursively.
+        3. Otherwise, add it to self.explorerFrame.expStructure.
+        """
+        expList = cmdtools.run_command(f"artiq_client ls {path}").stdout
+        expList = expList.split("\n")[:-1]  # The last one is always an empty string.
+        for expFile in expList:
+            if expFile.endswith("/") and not expFile.startswith("_"):
+                expFileItem = QTreeWidgetItem(parent)
+                expFileItem.setText(0, expFile[:-1])
```

Comment:
How about just leaving the trailing "/"?
It would be helpful to distinguish directories from files for the GUI users as well.

Reply:
Originally, I appended the "/" to directories, but it looks a little weird for me.
I thinks it is easy enough to distinguish because directories display toggles.
Nevertheless, do you think it can be confusing?

Reply:
Oh 😄 okay then. Actually it is not so confusing.